### PR TITLE
feat: New wellsLayer property: "simplifiedRendering".  Default false.

### PR DESCRIPTION
**Pull request by nilscb**
If enabled will cause wells to be drawn in simplified mode when rotaing or panning to increase drawing speed.
 Added new story for gullfaks wells to test large data sets.
 Also existing "refine" property defect was fixed.

### CODE REVIEW

File: typescript/packages/subsurface-viewer/src/components/Map.tsx [hunk 1]
```diff
@@ -324,7 +337,7 @@ export interface MapProps {
      * Each JSON object will consist of layer type with key as "@@type" and
      * layer specific data, if any.
      */
-    layers?: LayersList;
+    layers_list?: LayersList;
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/components/Map.tsx:340]: You can't simply break the API :/

Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/components/Map.tsx:340]: Good point! I will fix it

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 2]
```diff
@@ -97,6 +98,10 @@ export interface WellsLayerProps extends ExtendedLayerProps {
      * For example depth of z = 1000 corresponds to -1000 on the z axis. Default true.
      */
     ZIncreasingDownwards: boolean;
+    drawSimple: boolean;
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:101]: optimizedRendering ?
renderingHint (as an enum { QualityRendering, FastRendering } ?

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 3]
```diff
@@ -97,6 +98,10 @@ export interface WellsLayerProps extends ExtendedLayerProps {
      * For example depth of z = 1000 corresponds to -1000 on the z axis. Default true.
      */
     ZIncreasingDownwards: boolean;
+    optimizedRendering: boolean;
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:101]: Can you add some doc ?

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 4]
```diff
@@ -578,15 +611,19 @@ export default class WellsLayer extends CompositeLayer<WellsLayerProps> {
...
-            outline,
-            log_layer,
-            colors,
-            highlight,
-            highlightMultiWells,
-            selection_layer,
-            names,
-        ];
+        if (this.props.optimizedRendering && this.props.optimizedInteraction) {
+            return [colors] as unknown as LayersList;
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:615]: colors as layer list ?

Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:615]: Its not the colors but the list. That is not "colors" but "[colors]"

Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts]: Ok I see that lint no longer complains about that line (it did!)  so I removed the type cast

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts (hunk 3 above)
Comment (t0oF-azpn) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:101]: optimizedInteraction is documented, but not optimizedRendering

Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:101]: Doc is missing for `optimizedRendering`

Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts]: That one is for internal use (not to be set by user) but I put some documentation there.

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts (hunk 4 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts]: Agree on that: I havent changed existing names but should probably do it later.

File: typescript/packages/subsurface-viewer/src/components/Map.tsx [hunk 5]
```diff
@@ -579,6 +592,28 @@ const Map: React.FC<MapProps> = ({
...
+    React.useEffect(() => {
+        ZScaleOrbitController.setIsRotatingReference(setIsRotating);
+    }, [setIsRotating]);
+
+    useEffect(() => {
+        setIsRotState(isRotating);
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [isRotating]);
+
+    const layers_list = layers?.map((item) => {
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/components/Map.tsx:608]: I don't think we should add this handling in the Map component. It should be fairly straightforward to use [`onDragstart`](https://deck.gl/docs/api-reference/core/deck#ondragstart), etc. to manipulate layers' state.

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/components/Map.tsx (hunk 5 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/components/Map.tsx]: Done :-)

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts [hunk 6]
```diff
@@ -107,6 +107,13 @@ export function CatmullRom(
     return [x, y, z] as Position3D;
 }
 
+// Generate a range of numbers with given step length.
+const range = (start: number, stop: number, step: number) =>
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts:111]: Is this the same as lodash's [`range` method?](https://lodash.com/docs/#range)

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 7]
```diff
@@ -417,11 +487,12 @@ export default class WellsLayer extends CompositeLayer<WellsLayerProps> {
                 getFillColor: getColor(this.props.wellHeadStyle?.color),
                 getLineColor: getColor(this.props.lineStyle?.color),
                 parameters,
+                visible: this.props.logCurves && !drawSimple,
             })
         );
 
         // Highlight the multi selected wells.
-        const highlightMultiWells = new UnfoldedGeoJsonLayer(
+        const highlightMultiWells_layer = new UnfoldedGeoJsonLayer(
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:495]: Let's stick to one naming convention

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts (hunk 6 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts:111]: Good point

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts [hunk 8]
```diff
@@ -222,6 +228,64 @@ export function splineRefine(data_in: FeatureCollection): FeatureCollection {
...
+
+        for (let i = 1; i < n - 1; i += 1) {
+            if (i % 5) {
+                const P1 = coords[i + 0];
+                const md1 = mds[0][i + 0];
+
+                newCoordinates.push(P1);
+                newMds[0].push(md1);
+            }
+        }
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts:271]: Could you use [lodash's `remove`](https://lodash.com/docs/4.17.15#remove)?

Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts]: Good point. But I need the index of the elements so I used the  standard javascript filter method instead.

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 9]
```diff
@@ -434,6 +434,69 @@ Wells3d.parameters = {
...
+    const [optimizedInteraction, setOptimizedInteraction] =
+        React.useState<boolean>(false);
+
+    const props = {
+        ...args,
+        onDragStart: () => setOptimizedInteraction(true),
+        onDragEnd: () => setOptimizedInteraction(false),
+        layers: [
+            {
+                "@@type": "WellsLayer",
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:450]: I think it's better to use the `new` syntax for better type checking.

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 10]
```diff
@@ -97,6 +103,10 @@ export interface WellsLayerProps extends ExtendedLayerProps {
      * For example depth of z = 1000 corresponds to -1000 on the z axis. Default true.
      */
     ZIncreasingDownwards: boolean;
+    /**  If true means that a simplified representation of the wells will be drawn.
+     *   Useful for example during panning and roation to gain speed.
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:107]: roation => rotation

File: typescript/packages/subsurface-viewer/src/SubsurfaceViewer.tsx [hunk 11]
```diff
@@ -99,6 +99,9 @@ export interface SubsurfaceViewerProps {
      */
     isLoadedCallback?: (arg: boolean) => void;
 
+    onDragStart?: (info: unknown, event: unknown) => void;
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/SubsurfaceViewer.tsx:102]: We should try to type the arguments.
info is a PickingInfo, right ?
event is burried too deep to have really typing info :(

File: typescript/packages/subsurface-viewer/src/components/Map.tsx [hunk 12]
```diff
@@ -397,6 +397,9 @@ export interface MapProps {
      */
     isLoadedCallback?: (arg: boolean) => void;
 
+    onDragStart?: (info: unknown, event: unknown) => void;
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/components/Map.tsx:400]: Question: can't this be done using onMouseEvent ?
Or at lease, use the same approach: onDragStart?: (pickInfo: PickingInfo, event: unknown) => void;

I guess there should definitively be some cleaning of the events. Some are unknown, some are translated to MapMouseEvent (which should evolve to be more consistent across objects; it has special fields for wells :( )

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 13]
```diff
@@ -340,7 +381,35 @@ export default class WellsLayer extends CompositeLayer<WellsLayerProps> {
             [GL.POLYGON_OFFSET_FILL]: true,
         };
 
-        const outline = new UnfoldedGeoJsonLayer(
+        // Reduced details when rotating or panning the view if "optimizedInteraction" is set.
+        const drawSimple = this.props.optimizedInteraction;
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:385]: avoid using simple (it does not give a hint).
fastDrawing is more meaningful.

**Review by hkfb — COMMENTED** [review]
Should we consider using [@turf/simplify](https://turfjs.org/docs/#simplify) instead?

It uses the [Ramer-Douglas-Peucker algorithm](https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm) to simplify geojson geometries.

File: typescript/package.json [hunk 14]
```diff
@@ -152,5 +152,9 @@
     },
     "workspaces": [
         "packages/*"
-    ]
+    ],
+    "dependencies": {
+        "@turf/simplify": "^6.5.0",
+        "@turf/turf": "^6.5.0"
+    }
```
Comment (hkfb) [inline typescript/package.json:159]: This should be in the `package.json` of the `subsurface-viewer` package.

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 15]
```diff
@@ -434,6 +437,64 @@ Wells3d.parameters = {
     },
 };
 
+// Gullfaks wells.
+export const Wells3DGullfaks: ComponentStory<typeof SubsurfaceViewer> = (
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:441]: I think this story should have a more descriptive name, like "SimplifiedRendering"

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 16]
```diff
@@ -97,6 +103,10 @@ export interface WellsLayerProps extends ExtendedLayerProps {
      * For example depth of z = 1000 corresponds to -1000 on the z axis. Default true.
      */
     ZIncreasingDownwards: boolean;
+    /**  If true means that a simplified representation of the wells will be drawn.
+     *   Useful for example during panning and roation to gain speed.
+     */
+    optimizedInteraction: boolean;
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:109]: Agree with @t0oF-azpn, it's not the interaction that controls the rendering, it's the consumer using a callback, and "optimized" is not the right word. I think "simplifiedRendering" fits better

**Review by hkfb — CHANGES_REQUESTED** [review]

**Review by hkfb — APPROVED** [review]
LGTM :+1:

**Pull request merged**